### PR TITLE
update python sdk to the latest REST API

### PR DESCRIPTION
**Pull request by magnurud**

### CODE REVIEW

File: las/client.py [hunk 1]
```diff
@@ -126,236 +139,438 @@ def create_document(self, content: bytes, content_type: str, consent_id: str,
...
+        :param user_id: Id of the user
         :type user_id: str
-        :return: batch handle id and pre-signed upload url
+        :return: User response from REST API
         :rtype: dict
 
         :raises: :py:class:`~las.InvalidCredentialsException`, :py:class:`~las.TooManyRequestsException`,\
  :py:class:`~las.LimitExceededException`, :py:class:`requests.exception.RequestException`
         """
+        return self._make_request(requests.delete, f'/users/{user_id}', body={})
```
Comment (krigar1184) [inline las/client.py:429]: DELETE requests should not have body.

File: las/client.py [hunk 2]
```diff
@@ -126,236 +139,438 @@ def create_document(self, content: bytes, content_type: str, consent_id: str,
...
+
+        :param workflow_id: Id of the workflow
+        :type workflow_id: str
+        :return: Workflow response from REST API
+        :rtype: dict
+
+        :raises: :py:class:`~las.InvalidCredentialsException`, :py:class:`~las.TooManyRequestsException`,\
+ :py:class:`~las.LimitExceededException`, :py:class:`requests.exception.RequestException`
+        """
+        return self._make_request(requests.delete, f'/workflows/{workflow_id}', body={})
```
Comment (krigar1184) [inline las/client.py:494]: Same as above.

File: las/client.py [hunk 3]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
             'batchId': batch_id,
-            'feedback': feedback
+            'groundTruth': ground_truth,
         }
-        return self._make_request(requests.post, '/documents', body=dictstrip(body), encode_body=False)
+        return self._make_request(requests.post, '/documents', body=dictstrip(body))
 
-    def list_documents(self, batch_id: Optional[str] = None, consent_id: Optional[str] = None) -> dict:
-        """List documents that you have created, calls the GET /documents endpoint.
+    def list_documents(self, batch_id: Optional[Queryparam] = None, consent_id: Optional[Queryparam] = None,
```
Comment (augustak) [inline las/client.py:246]: arguments on separate lines

File: las/client.py [hunk 4]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+        :rtype: dict
 
-        return self._make_request(requests.get, '/documents', params={'batchId': batch_id, 'consentId': consent_id})
+        :raises: :py:class:`~las.InvalidCredentialsException`, :py:class:`~las.TooManyRequestsException`,\
+ :py:class:`~las.LimitExceededException`, :py:class:`requests.exception.RequestException`
+        """
+        return self._make_request(requests.get, f'/documents/{document_id}')
 
-    def create_prediction(self, document_id: str, model_name: str, max_pages: Optional[int] = None,
+    def update_document(self, document_id: str, ground_truth: Sequence[Dict[str, Union[Optional[str], bool]]]) -> dict:
```
Comment (augustak) [inline las/client.py:310]: Does it make sense to use the Field object here?

File: las/client.py [hunk 5]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+        :type ground_truth: Sequence[Dict[str, Union[Optional[str], bool]]]
+        :return: Document response from REST API
+        :rtype: dict
+
+        :raises: :py:class:`~las.InvalidCredentialsException`, :py:class:`~las.TooManyRequestsException`,\
+ :py:class:`~las.LimitExceededException`, :py:class:`requests.exception.RequestException`
+        """
+        return self._make_request(requests.patch, f'/documents/{document_id}', body={'groundTruth': ground_truth})
+
+    def create_prediction(self, document_id: str, model_id: str, max_pages: Optional[int] = None,
```
Comment (augustak) [inline las/client.py:332]: Arguments on separate lines

File: las/client.py [hunk 6]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
                           auto_rotate: Optional[bool] = None, extras: Dict[str, Any] = None) -> dict:
```
Comment (augustak) [inline las/client.py:333]: extras can be removed for now

File: las/client.py [hunk 7]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+        """
+        body = {
+            'data': data,
+            'description': description,
+        }
+        return self._make_request(requests.patch, f'/secrets/{secret_id}', body=dictstrip(body))
 
-    def delete_consent(self, consent_id: str) -> dict:
-        """Delete documents with this consent_id, calls the DELETE /consents/{consentId} endpoint.
+    def create_transition(self, name, transition_type: str, in_schema: dict, out_schema: dict,
```
Comment (augustak) [inline las/client.py:439]: Arguments on separate lines

File: las/client.py [hunk 8]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+
+        :param in_schema: Json-schema that defines the input to the transition
+        :type in_schema: dict
+        :param out_schema: Json-schema that defines the output of the transition
+        :type out_schema: dict
+        :param name: Name of the transition
+        :type name: str
+        :param transition_type: Type of transition "docker"|"manual"
+        :type transition_type: str
+        :param params: Extra parameters to the transition
```
Comment (augustak) [inline las/client.py:465]: Does 'Extra' sounds too generic? "Parameters to corresponding transition_type." + An overview of the supported parameters by transition_type?

File: las/client.py [hunk 9]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+            'description': description,
+            'inputJsonSchema': in_schema,
+            'outputJsonSchema': out_schema,
+            'transitionType': transition_type,
+            'params': params,
+        }
+        return self._make_request(requests.post, '/transitions', body=dictstrip(body))
+
+    def list_transitions(self, transition_type: Optional[Queryparam] = None,
+                         max_results: Optional[int] = None, next_token: Optional[str] = None) -> dict:
```
Comment (augustak) [inline las/client.py:486]: Arguments on separate lines

File: las/client.py [hunk 10]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+        params = {
+            'transitionType': transition_type,
+            'maxResults': max_results,
+            'nextToken': next_token,
+        }
+        return self._make_request(requests.get, url, params=params)
 
-    def create_batch(self, description: str) -> dict:
-        """Creates a batch handle, calls the POST /batches endpoint.
+    def update_transition(self, transition_id: str, *, name: Optional[str], in_schema: Optional[dict],
```
Comment (augustak) [inline las/client.py:513]: Arguments on separate lines

File: las/client.py [hunk 11]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
-        return self._make_request(requests.post, '/batches', body={'description': description})
+        endpoint = f'/transitions/{transition_id}/executions'
+        return self._make_request(requests.post, endpoint, body={})
+
+    def list_transition_executions(
+            self, transition_id: str,
+            status: Optional[Queryparam] = None,
+            execution_id: Optional[Queryparam] = None,
+            max_results: Optional[int] = None,
+            next_token: Optional[str] = None
```
Comment (augustak) [inline las/client.py:570]: trailing comma

File: las/client.py [hunk 12]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+        params = {
+            'status': status,
+            'executionId': execution_id,
+            'maxResults': max_results,
+            'nextToken': next_token,
+        }
+        return self._make_request(requests.get, url, params=params)
+
+    def update_transition_execution(self, transition_id: str, execution_id: str, status: str,
+                                    output: Optional[dict] = None, error: Optional[dict] = None) -> dict:
```
Comment (augustak) [inline las/client.py:604]: Arguments on separate lines

File: las/client.py [hunk 13]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+        """Creates a new workflow, calls the POST /workflows endpoint.
+
+        >>> from las.client import BaseClient
+        >>> from pathlib import Path
+        >>> client = BaseClient()
+        >>> specification = {'language': 'ASL', 'version': '1.0.0', 'definition': {...}}
+        >>> error_config = {'email': '<error-recipient>'}
+        >>> client.create_workflow(specification, '<name>', '<description>', error_config)
+
+        :param specification: Specification of the workflow
```
Comment (augustak) [inline las/client.py:724]: Would a link to https://states-language.net/spec.html make sense? + The limitations imposed by our API (can add later)

File: las/client.py [hunk 14]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+        :raises: :py:class:`~las.InvalidCredentialsException`, :py:class:`~las.TooManyRequestsException`,\
+ :py:class:`~las.LimitExceededException`, :py:class:`requests.exception.RequestException`
+        """
+        params = {
+            'maxResults': max_results,
+            'nextToken': next_token,
+        }
+        return self._make_request(requests.get, '/workflows', params=params)
+
+    def update_workflow(self, workflow_id: str, *, name: Optional[str], description: Optional[str] = None) -> dict:
```
Comment (augustak) [inline las/client.py:769]: Using '*' syntax here but not in most of the functions? I am in favor of using '*' for all optional arguments

File: las/client.py [hunk 15]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+        return self._make_request(requests.post, endpoint, body={'input': content})
+
+    def list_workflow_executions(
+            self,
+            workflow_id: str,
+            status: Optional[Queryparam] = None,
+            sort_by: Optional[str] = None,
+            order: Optional[str] = None,
+            max_results: Optional[int] = None,
+            next_token: Optional[str] = None
```
Comment (augustak) [inline las/client.py:842]: trailing comma

File: las/client.py [hunk 16]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+        """
+        body = {
+            'name': name,
+            'description': description,
+            'inputJsonSchema': in_schema,
+            'outputJsonSchema': out_schema,
+        }
+        return self._make_request(requests.patch, f'/transitions/{transition_id}', body=dictstrip(body))
+
+    def execute_transition(self, transition_id: str) -> dict:
```
[the remark this answers is not in the document]
Comment (magnurud) [reply las/client.py:547]: it seems very verbose, are you sure?

File: las/client.py [hunk 17]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
-        return Prediction(document_id, consent_id, model_name, prediction_response)
+        prediction_response = self.create_prediction(document_id, model_id, extras=extras)
+        return Prediction(document_id, consent_id, model_id, prediction_response)
 
-    def send_feedback(self, document_id: str, feedback: List[Field]) -> dict:
-        """Send feedback to the model.
-        This method takes care of sending feedback related to document specified by document_id.
-        Feedback consists of ground truth values for the document specified as a list of\
- :py:class:`~las.Field` instances.
+    def send_ground_truth(self, document_id: str, ground_truth: List[Field]) -> dict:
```
[the remark this answers is not in the document]
Comment (magnurud) [reply las/client.py:932]: To accompany the predict function. It also makes the update_document method more accessible. Right now it is not obvious what update_document does based on the name alone, while send_ground_truth is easier to grasp.

File: las/client.py (hunk 4 above)
Comment (magnurud) [reply las/client.py:310]: This is also a difference between the client and the base client which I believe is worth to conserve.

File: las/prediction.py [hunk 18]
```diff
@@ -21,11 +21,11 @@ def confidence(self):
 
 
 class Prediction(Dict[str, str]):
```
Comment (krigar1184) [inline las/prediction.py:23]: I know this is old code, but inheriting from dict doesn't make any sense, prediction doesn't have anything to do with it.

File: las/client.py [hunk 19]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
         """
-
         content_type = self._get_content_type(document_path)
         document = pathlib.Path(document_path).read_bytes()
         response = self.create_document(document, content_type, consent_id)
         document_id = response['documentId']
-        prediction_response = self.create_prediction(document_id, model_name)
-        return Prediction(document_id, consent_id, model_name, prediction_response)
+        prediction_response = self.create_prediction(document_id, model_id, extras=extras)
+        return Prediction(document_id, consent_id, model_id, prediction_response)
```
Comment (krigar1184) [inline las/client.py:1004]: `consent` has type `Optional[str`], but `Prediction` accepts `str`.

File: las/prediction.py (hunk 18 above)
Comment (krigar1184) [reply las/prediction.py:23]: Same applies to `Field`.

File: tests/test_config.cfg [hunk 20]
```diff
@@ -1,4 +1,4 @@
 [default]
-model_names = invoice,receipt
+model_ids = las:model:014f1a205b4443e9ad1ea524ac963a3e, las:model:44deb3ea25aa443b852b0c8f96bcb4ee
```
Comment (krigar1184) [inline tests/test_config.cfg:2]: What are these values?

File: tests/util.py [hunk 21]
```diff
@@ -0,0 +1,52 @@
+from uuid import uuid4
```
Comment (krigar1184) [inline tests/util.py:1]: These should be pytest fixtures.

**Review by krigar1184 — CHANGES_REQUESTED** [review]

File: tests/test_config.cfg (hunk 20 above)
Comment (magnurud) [reply tests/test_config.cfg:2]: just test values... This is an old design, maybe time to replace it with fixtures?

File: las/client.py [hunk 22]
```diff
@@ -89,303 +90,871 @@ def _create_signing_headers(self, path: str):
...
+        :type workflow_id: str
+        :return: Workflow response from REST API
+        :rtype: dict
+
+        :raises: :py:class:`~las.InvalidCredentialsException`, :py:class:`~las.TooManyRequestsException`,\
+ :py:class:`~las.LimitExceededException`, :py:class:`requests.exception.RequestException`
+        """
+        return self._make_request(requests.delete, f'/workflows/{workflow_id}')
+
+    def execute_workflow(self, workflow_id: str, content: dict) -> dict:
```
[the remark this answers is not in the document]
Comment (augustak) [reply las/client.py]: My OCD is sufferering, but I can go along with that

File: tests/util.py (hunk 21 above)
Comment (magnurud) [reply tests/util.py:1]: How would you solve this for instance? 
```
@pytest.mark.parametrize('transition_type,params', [
    ('docker', {'imageUrl': 'python3.8'}),
    ('manual', None),
    ('manual', {'assets': {'jsRemoteComponent': util.asset_id()}}),
])
def test_create_transition(client: BaseClient, transition_type, params, schema):
```

File: las/client.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (augustak) [reply las/client.py]: I don't think we want two functions doing basically the same thing. I think updating a document with ground truth is pretty self explanatory. Also, we should have general guidelines / docs that explain these interactions anyway.

File: tests/util.py (hunk 21 above)
Comment (krigar1184) [reply tests/util.py:1]: Things connected to domain logic don't belong in utility modules. A simpler solution would be moving them to another module and prefixing them with `create_`.

**Review by krigar1184 — APPROVED** [review]

File: las/client.py [hunk 23]
```diff
@@ -76,322 +74,891 @@ class FileFormatException(ClientException):
...
-        }
-        headers = {**auth_headers, 'Content-Type': 'application/json'}
-        return uri, headers
-
     @on_exception(expo, TooManyRequestsException, max_tries=4)
     @on_exception(expo, RequestException, max_tries=3, giveup=_fatal_code)
     def _make_request(self, requests_fn: Callable, signing_path: str,
-                      body: Optional[dict] = None, params: Optional[dict] = None,
-                      encode_body: bool = True) -> dict:
+                      body: Optional[dict] = None, params: Optional[dict] = None) -> Dict:
```
Comment (augustak) [inline las/client.py:88]: arguments on separate lines

**Pull request merged**